plugins/sqlfluff-templater-dbt/sqlfluff_templater_dbt/templater.py

```diff
@@ -70,6 +72,103 @@ class DbtConfigArgs:
     REQUIRE_RESOURCE_NAMES_WITHOUT_SPACES: Optional[bool] = None
 
 
+def is_dbt_exception(exception: Optional[BaseException]) -> bool:
+    """Check whether this looks like a dbt exception."""
+    # None is not a dbt exception.
+    if not exception:
+        return False
+    return exception.__class__.__module__.startswith("dbt")
+
+
+def _extract_error_detail(exception: BaseException) -> str:
+    """Serialise an exception into a string for reuse in other messages."""
+    return (
+        f"{exception.__class__.__module__}.{exception.__class__.__name__}: {exception}"
+    )
+
+
+T = TypeVar("T")
+
+
+def handle_dbt_errors(
+    error_class: Type[Exception], preamble: str
+) -> Callable[[Callable[..., T]], Callable[..., T]]:
+    """A decorator to safely catch dbt exceptions and raise native ones.
+
+    NOTE: This looks and behaves a lot like a context manager, but it's
+    important that it is *not* a context manager so that it can effectively
+    strip the context from handled exceptions. That isn't possible (as far
+    as we've tried) within a context manager.
+
+    dbt exceptions don't pickle nicely, and python exception context tries
+    very hard to make sure that the exception context of any new exceptions
+    is preserved. This means we have to be quite deliberate in stripping any
+    dbt exceptions, not just those that are directly raised, but those which
+    are present within the `__context__` or `__cause__` attributes of any
+    SQLFluff exceptions.
+
+    This wrapper aims to do that, catching any dbt exceptions and
+    raising SQLFluff exceptions, and also making sure that any native
+    SQLFluff exceptions which are handled are also stripped of any
+    unwanted dbt exceptions so that we don't cause issues when in
+    multithreaded/multiprocess operation.
+
+    https://docs.python.org/3/library/exceptions.html#inheriting-from-built-in-exceptions
+    https://github.com/sqlfluff/sqlfluff/issues/6037
+    """  # noqa E501
+
+    def decorator(func: Callable[..., T]) -> Callable[..., T]:
+        def wrapped_method(*args, **kwargs) -> T:
+            # NOTE: `_detail` also acts as a semaphore to indicate whether an exception
```

Comment:
Nit: Can we say "flag" rather than "semaphore" here? In Python, there's an actual class, [Semaphore](https://docs.python.org/3/library/threading.html#semaphore-objects), that is used in parallel programming, so the term could be confusing here, given that this decorator helps with dbt exception passing in parallel programming.

Reply:
Good idea 👍